src/charm.py

```diff
@@ -0,0 +1,178 @@
+
+        return S3ConnectionInfo(
+            **{key.replace("-", "_"): value for key, value in raw.items() if key != "data"}
+        )
+
+    def get_status(
+        self,
+        s3: Optional[S3ConnectionInfo],
+    ) -> StatusBase:
+        """Compute and return the status of the charm."""
+        if not self.k8s_permissions:
+            return Status.MISSING_PERMISSIONS.value
+
+        if not s3:
+            return Status.MISSING_S3_RELATION.value
+
+        if not s3.verify():
+            return Status.INVALID_CREDENTIALS.value
+
+        return Status.ACTIVE.value
+
+    def update_service(self, s3: Optional[S3ConnectionInfo]) -> bool:
+        """Update the Spark Configuration hub service if needed."""
+        status = self.log_result(lambda _: f"Status: {_}")(self.get_status(s3))
+
+        self.unit.status = status
+        spark_config = SparkConfigurationHubConfig(s3)
+        self.logger.info(f"Current configuration: {spark_config.contents}")
+        # fid.write(spark_config.contents)
+
+        with open(CONF_FILE_PATH, "w") as f:
+            f.write(spark_config.contents)
+            f.close()
+
+        if status is not Status.ACTIVE.value:
+            self.logger.info(f"Cannot start service because of status {status}")
+            return False
+        if pid := self.get_peer_data("pid"):
+            self._stop_process(int(pid))
+        self._start_process()
+        return True
+
+    def _on_install(self, _: InstallEvent) -> None:
+        """Handle the `on_install` event."""
+        self.unit.status = Status.INSTALL.value
+
+    def _on_s3_credential_changed(self, _: CredentialsChangedEvent):
+        """Handle the `CredentialsChangedEvent` event from S3 integrator."""
+        self.logger.info("S3 Credentials changed")
+        self.update_service(self.s3_connection_info)
+
+    def _on_s3_credential_gone(self, _: CredentialsGoneEvent):
+        """Handle the `CredentialsGoneEvent` event for S3 integrator."""
+        self.logger.info("S3 Credentials gone")
+        self.update_service(None)
+
+    def _update_event(self, _):
+        """Handle the update event hook."""
+        self.unit.status = self.get_status(self.s3_connection_info)
+
+    def _on_config_changed(self, event: ConfigChangedEvent):
+        """Handle the on config changed event."""
+        self.update_service(self.s3_connection_info)
+
+    def _start_process(
+        self,
+    ) -> int:
+        """Handle the start of the process for consumption or production of messagges."""
+        t0 = int(time())
+        my_cmd = (
+            f"nohup python3 -m src.monitor_sa --app-name={self.app.name} --config={CONF_FILE_PATH}"
+        )
+        self.logger.info(my_cmd)
+        process = subprocess.Popen(
+            shlex.split(my_cmd),
+            stdout=open(f"/tmp/{t0}_monitor_sa.log", "w"),
+            stderr=open(f"/tmp/{t0}_monitor_sa.err", "w"),
+        )
+        self.set_peer_data(key="pid", data=str(process.pid))
```

Comment:
**issue** Here there is an issue because we are writing this to the app databag, but we should probably write this to the unit databag since it should be relevant for the unit. But also, after some thinking, I'm actually not sure we should write this to the unit databag either. If the pod were to be rescheduled, this information would still be present, but probably with no meaning (since there is no process with that pid). 

What about writing this information to a local file in the container instead? I kinda like this, because in this way we don't even need to provide the workload with charm information. In the new pattern, workload should be independent from charm related objects, and indeed this is also suggested by the reasoning above. 

ps. If I had the pattern when writing the kafka-test-app, I would probably avoided this oversight there as well